pkg/api/applications/v2/http.go

```diff
@@ -0,0 +1,556 @@
+/*
+Copyright 2021 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/thestormforge/optimize-go/pkg/api"
+)
+
+const endpointApplications = "/v2/applications/"
+
+func NewAPI(c api.Client) API {
+	return &httpAPI{client: c}
+}
+
+type httpAPI struct {
+	client api.Client
+}
+
+var _ API = &httpAPI{}
+
+func (h *httpAPI) ListApplications(ctx context.Context, q ApplicationListQuery) (ApplicationList, error) {
+	u := h.client.URL(endpointApplications)
+	u.RawQuery = url.Values(q.IndexQuery).Encode()
+
+	return h.ListApplicationsByPage(ctx, u.String())
+}
+
+func (h *httpAPI) ListApplicationsByPage(ctx context.Context, u string) (ApplicationList, error) {
+	result := ApplicationList{}
+
+	req, err := http.NewRequest(http.MethodGet, u, nil)
+	if err != nil {
+		return result, err
+	}
+
+	resp, body, err := h.client.Do(ctx, req)
+	if err != nil {
+		return result, err
+	}
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		result.Metadata = api.Metadata(resp.Header)
+		err = json.Unmarshal(body, &result)
+		return result, err
+	default:
+		return result, api.NewUnexpectedError(resp, body)
+	}
+}
+
+func (h *httpAPI) CreateApplication(ctx context.Context, app Application) (api.Metadata, error) {
+	u := h.client.URL(endpointApplications).String()
+
+	req, err := httpNewJSONRequest(http.MethodPost, u, app)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, body, err := h.client.Do(ctx, req)
```

Comment:
This almost got me with the similarities to net/http client.Do. Was worried we were leaking body for `http.StatusCreated`, but in api.Client body is just `[]byte`

Reply:
I shamelessly took inspiration from the [Prometheus Go client](https://github.com/prometheus/client_golang/blob/master/api/client.go) many moons ago.